sensing/src/test/java/com/google/android/sensing/impl/SensorManagerImplTest.kt

```diff
@@ -0,0 +1,227 @@
+    val testLifecycleOwner = TestLifecycleOwner()
+
+    runTest {
+      sensorManager.init(TEST_SENSOR_TYPE, mockContext, testLifecycleOwner, TestSensorInitConfig())
+      sensorManager.start(TEST_SENSOR_TYPE, TestSensorCaptureRequest())
+      sensorManager.reset(TEST_SENSOR_TYPE)
+
+      // start will give IllegalStateException
+      try {
+        sensorManager.start(TEST_SENSOR_TYPE, TestSensorCaptureRequest())
+      } catch (e: Exception) {
+        assert(e is java.lang.IllegalStateException)
+      }
+    }
+  }
+
+  @Test
+  fun registerSensorFactory_registerListenerThenStart_shouldReceiveCaptureEvents() {
+    sensorManager.registerSensorFactory(TEST_SENSOR_TYPE, TEST_SENSOR_FACTORY)
+    val mockContext = Mockito.mock(Context::class.java)
+    val testLifecycleOwner = TestLifecycleOwner()
+
+    runTest {
+      sensorManager.init(TEST_SENSOR_TYPE, mockContext, testLifecycleOwner, TestSensorInitConfig())
+      sensorManager.registerListener(
+        TEST_SENSOR_TYPE,
+        object : SensorManager.AppDataCaptureListener {
+          override fun onStart(captureInfo: CaptureInfo) {
+            assert(captureInfo.captureFolder == TEST_OUTPUT_FOLDER)
+            assert(captureInfo.externalIdentifier == TEST_EXTERNAL_ID)
+          }
+
+          override fun onComplete(captureInfo: CaptureInfo) {
+            assert(captureInfo.resourceInfoList.isNotEmpty())
+          }
+
+          override fun onError(exception: Exception, captureInfo: CaptureInfo?) {}
+        }
+      )
+      sensorManager.start(TEST_SENSOR_TYPE, TestSensorCaptureRequest())
+    }
+  }
+
+  @Test
+  fun registerSensorFactoryListenerThenStart_stopCapture_shouldCreateDbRecords() {
+    sensorManager.registerSensorFactory(TEST_SENSOR_TYPE, TEST_SENSOR_FACTORY)
+    val mockContext = Mockito.mock(Context::class.java)
+    val testLifecycleOwner = TestLifecycleOwner()
+
+    runTest {
+      sensorManager.init(TEST_SENSOR_TYPE, mockContext, testLifecycleOwner, TestSensorInitConfig())
+      sensorManager.registerListener(
+        TEST_SENSOR_TYPE,
+        object : SensorManager.AppDataCaptureListener {
+          override fun onStart(captureInfo: CaptureInfo) {
+            assert(captureInfo.captureId != null)
+            assert(captureInfo.captureFolder == TEST_OUTPUT_FOLDER)
+            assert(captureInfo.externalIdentifier == TEST_EXTERNAL_ID)
+          }
+
+          override fun onComplete(captureInfo: CaptureInfo) {
+            assert(captureInfo.captureId != null)
+            assert(captureInfo.resourceInfoList.isNotEmpty())
+            runTest {
+              val captureInfoInRecords = sensingEngine.getCaptureInfo(captureInfo.captureId!!)
+              assert(captureInfoInRecords == captureInfo)
+            }
+          }
+
+          override fun onError(exception: Exception, captureInfo: CaptureInfo?) {}
+        }
+      )
+      sensorManager.start(TEST_SENSOR_TYPE, TestSensorCaptureRequest())
```

Comment:
You could also assert if the `onStart` and `onComplete` methods are called or not.

Reply:
Due to this issue I am not able to implement this. Since the app callback is invoked in another job, test functions are not able to track it. 

This is already solved in the hear branch. This is the PR #98 which makes app callback faster.

Reply:
Created [this issue](https://github.com/google-research/CVD-paper-mobile-camera-example/issues/109) to add these tests once the above issue is resolved.

Reply:
Ok @MJ1998 .